toolchain/parser/parser.cpp

```diff
@@ -1676,4 +1681,72 @@ auto Parser::HandleVarFinishAsNoSemicolonState() -> void {
   HandleVarFinish(/*require_semicolon=*/false);
 }
 
+auto Parser::HandleInterfaceIntroducerState() -> void {
+  auto state = PopState();
+
+  if (!ConsumeAndAddLeafNodeIf(TokenKind::Identifier(),
+                               ParseNodeKind::DeclaredName())) {
+    CARBON_DIAGNOSTIC(ExpectedInterfaceName, Error,
+                      "Expected interface name after `interface` keyword.");
+    emitter_.Emit(*position_, ExpectedInterfaceName);
+    state.has_error = true;
+  }
+
+  bool has_open_curly_brace = true;
+
+  if (!PositionIs(TokenKind::OpenCurlyBrace())) {
+    CARBON_DIAGNOSTIC(ExpectedInterfaceOpenCurlyBrace, Error,
+                      "Expected `{{` to start interface definition.");
+    emitter_.Emit(*position_, ExpectedInterfaceOpenCurlyBrace);
+    state.has_error = true;
+
+    if (auto next_open_curly_brace = FindNextOf({TokenKind::OpenCurlyBrace()});
+        next_open_curly_brace) {
+      SkipTo(*next_open_curly_brace);
+    } else {
+      has_open_curly_brace = false;
+    }
+  }
+
+  state.state = ParserState::InterfaceDefinitionFinish();
+  PushState(state);
+
+  if (has_open_curly_brace) {
+    PushState(ParserState::InterfaceDefinitionLoop());
+    ++position_;
+  }
+}
+
+auto Parser::HandleInterfaceDefinitionLoopState() -> void {
+  switch (PositionKind()) {
+    case TokenKind::CloseCurlyBrace(): {
+      AddLeafNode(ParseNodeKind::InterfaceBodyEnd(), *position_);
+      ++position_;
+
+      auto open_curly_state = PopState();
+      AddNode(ParseNodeKind::InterfaceBody(), open_curly_state.token,
+              open_curly_state.subtree_start, open_curly_state.has_error);
+
+      break;
+    }
+      // TODO Handle possible declarations inside interface body.
+    default: {
+      CARBON_DIAGNOSTIC(UnrecognizedDeclaration, Error,
+                        "Unrecognized declaration introducer.");
+      emitter_.Emit(*position_, UnrecognizedDeclaration);
+      tree_.has_errors_ = true;
+      if (auto semi = SkipPastLikelyEnd(*position_)) {
+        AddLeafNode(ParseNodeKind::EmptyDeclaration(), *semi,
+                    /*has_error=*/true);
+      }
```

Comment:
```suggestion
      if (auto semi = SkipPastLikelyEnd(*position_)) {
        AddLeafNode(ParseNodeKind::EmptyDeclaration(), *semi,
                    /*has_error=*/true);
      } else {
        ReturnErrorOnState();
      }
```

I want to avoid directly touching `tree_.has_errors_` (reminds me I should refactor DeclarationLoop too). Here in particular, we can note the error on the leaf if it's found; ReturnErrorOnState will add it to the current loop state so you should see the interface body indicated as having an error. In either case, we only indicate the error once and the interface may be _mostly_ usable if it's generally structurally valid (but equally, this may invalidate the interface subtree).

Reply:
https://github.com/carbon-language/carbon-lang/pull/2413 is what I ended up with in DeclarationLoop